pkg/cmd/auth/login/login.go

```
@@ -0,0 +1,284 @@
+		Long: heredoc.Doc(`Authenticate with a GitHub host.
+
+			This interactive command initializes your authentication state either by helping you log into
+			GitHub via browser-based OAuth or by accepting a Personal Access Token.
+
+			The interactivity can be avoided by specifying --with-token and passing a token on STDIN.
+		`),
+		Example: heredoc.Doc(`
+			$ gh auth login
+			# => do an interactive setup
+
+			$ gh auth login --with-token < mytoken.txt
+			# => read token from mytoken.txt and authenticate against github.com
+
+			$ gh auth login --hostname enterprise.internal --with-token < mytoken.txt
+			# => read token from mytoken.txt and authenticate against a GitHub Enterprise instance
+		`),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			isTTY := opts.IO.IsStdinTTY()
+
+			if !isTTY {
+				if !cmd.Flags().Changed("with-token") {
+					return &cmdutil.FlagError{Err: errors.New("--with-token required when not attached to tty")}
+				}
+			}
+
+			wt, _ := cmd.Flags().GetBool("with-token")
+			if wt {
+				if isTTY {
+					return &cmdutil.FlagError{Err: errors.New("expected token on STDIN")}
+				}
+				defer opts.IO.In.Close()
+				token, err := ioutil.ReadAll(opts.IO.In)
+				if err != nil {
+					return &cmdutil.FlagError{Err: fmt.Errorf("failed to read token from STDIN: %w", err)}
+				}
+
+				opts.Token = strings.TrimSpace(string(token))
+				if opts.Hostname == "" {
+					opts.Hostname = defaultHostname
+				}
+			}
+
+			if runF != nil {
+				return runF(opts)
+			}
+
+			return loginRun(opts)
+		},
+	}
+
+	cmd.Flags().StringVar(&opts.Hostname, "hostname", "", "The hostname of the GitHub instance to authenticate with")
+	cmd.Flags().Bool("with-token", false, "If specified, token is read from STDIN")
+
+	return cmd
+}
+
+func loginRun(opts *LoginOptions) error {
+	cfg, err := opts.Config()
+	if err != nil {
+		return err
+	}
+
+	if opts.Token != "" {
+		if opts.Hostname == "" {
+			panic("empty hostname would leak oauth_token")
```

This is nice precaution! Could we return an error in this case?

---

we could; but this felt more like something that could come up in development as opposed to a user's running code (and the panic is to prevent it from getting out of development).

I guess returning an error would have the same effect and there's no reason not to.